### PR TITLE
89 ai init

### LINKED ISSUE #113: AI Classes: State Machine, State. Singleton Pattern
Initialise the AI Classes needed. This is just a start point for the AI engine. A FSM, abstract State class, and a singleton class to apply the singleton pattern

Addresses
#89 

##Classes to make :
- [ ] Finite State Machine class
- [ ] Abstract State class
- [ ] Singleton class

### PR DESCRIPTION
**Pull request by Idraliski**
## What problem does this pull request address?
We need a start point of the AI engine, it needs a component for the ECS and be able to read the AI scripts.

## How does this pull request solve the problem?
This creates a couple of needed classes for the AI engine, The FSM (state machine) and NPC.
Implementation has been made through the ECS, Scene and System classes (Not dumped in EntryPoint).
Modified Cube.ua and exitCube.lua as test subjects to act as NPCs. As well as added States.lua which carries all states, this is being used by the 2 objects mentioned.
DeltaTime has been passed to Scene to slow down AI Update calls, so its not called every frame.

Created an NPC class to hold the FSM and will be used to implement the movement functions that will be exposed to Lua. A lot of functionality for NPC is not there currently, this will be addressed in the next PR

Creates a Finite State Machine class. Deals with its states and has an update method to be called during game loop. Working Update function to run Lua scripts to call the correct state's update function within Lua

closes #113, and addresses issue #89

image
image
image

## What testing has been performed?
- [x] Runs an AI script through Scene and Systems. Prints from AI scripts and displays on terminal.
- [x] Update AI call frequency modifiable, reduced to once every 2 seconds (roughly, based on deltaTime of frames)
- [x] Altered 2 objects to have different FSM with different running states. 
- [x] Builds successfully
- [x] Runs successfully

### CODE REVIEW

File: Bottlebrush_Engine/Bottlebrush_Engine/BBAI/NPC.h
Comment (The-Spaghetti-Is-Ready) [inline Bottlebrush_Engine/Bottlebrush_Engine/BBAI/NPC.h:1]: does the FSM for this class need to be a pointer?

Comment (Idraliski) [reply Bottlebrush_Engine/Bottlebrush_Engine/BBAI/NPC.h:1]: this is incomplete, its as a pointer currently as I am attempting to use the Singleton class for FSM

Comment (Idraliski) [reply Bottlebrush_Engine/Bottlebrush_Engine/BBAI/NPC.h:1]: changed it now, its just a member var now

**Review by Minywire — DISMISSED** [review]
This looks really good!

**Review by The-Spaghetti-Is-Ready — CHANGES_REQUESTED** [review]
As per Andrew's recommendations, changes are needed to this branch

**Comment by Idraliski** [comment]
Addressed issues brought up, please re-review

**Review by The-Spaghetti-Is-Ready — APPROVED** [review]
Looks good, I can't see any issues with this.  Thanks for adding delta time to scene as well, I completely forgot about that

**Review by Minywire — APPROVED** [review]
after looking at the conversation in discord and here i'm happy to approve this

**Pull request merged**